lib/semantics/check-io.cc

```diff
@@ -15,12 +15,62 @@
 #include "check-io.h"
 #include "expression.h"
 #include "tools.h"
+#include "../common/format.h"
 #include "../parser/tools.h"
 
 namespace Fortran::semantics {
 
 // TODO: C1234, C1235 -- defined I/O constraints
 
+void IoChecker::Enter(
+    const parser::Statement<common::Indirection<parser::FormatStmt>> &stmt) {
+  if (!stmt.label.has_value()) {
```

Comment:
Does resolve-labels check that format statements are labeled?  If not, should this check be moved there?

Reply:
No, resolve-labels does not currently check that a format statement has a label.

I can move the subset of this function that does this constraint check to a Pre function in resolve-labels.cc, but doing that seems to silently suppress the call to this new function, suppressing the rest of the checks in this new function.  This behavior may be related to the comment near the beginning of semantics.cc -- _No two checkers may have the same Enter or Leave function._

Reply:
An unlabeled `FORMAT` statement is irrelevant to the program and perhaps shouldn't even be validated.

Ditto for unreferenced labeled `FORMAT`, although we can't figure that out completely yet until we can handle `ASSIGN`.

Reply:
> I can move the subset of this function that does this constraint check to a Pre function in resolve-labels.cc, but doing that seems to silently suppress the call to this new function, suppressing the rest of the checks in this new function.

If `ValidateLabels()` detects an error, semantic checking stops. This is because phases after that depend on label DO statements being rewritten to DO constructs. For that reason it would be better to leave this check outside of `resolve-labels.cc`.